src/go/util/httppattern/match_sequence_generator_node.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httppattern
+
+import (
+	"sort"
+)
+
+const HttpMethodWildCard = "*"
+
+// Trie node to implement match uence generator.
+type matchSequenceGeneratorNode struct {
+	ResultMap map[string]*lookupResult
+	Children  map[string]*matchSequenceGeneratorNode
+	WildCard  bool
+}
+
+func newMatchSequenceGeneratorNode() *matchSequenceGeneratorNode {
+	return &matchSequenceGeneratorNode{
+		ResultMap: make(map[string]*lookupResult),
+		Children:  make(map[string]*matchSequenceGeneratorNode),
+	}
+}
+
+func (hn *matchSequenceGeneratorNode) insertTemplate(pathParts []string, pathPartsIdxCur int, httpMethod string, methodData *methodData, markDuplicate bool) bool {
+	if pathPartsIdxCur == len(pathParts) {
+		if val, ok := hn.ResultMap[httpMethod]; ok {
+			if markDuplicate {
+				val.isMultiple = true
+			}
+			return false
+		}
+		hn.ResultMap[httpMethod] = &lookupResult{
+			data:       methodData,
+			isMultiple: false,
+		}
+		return true
+	}
+
+	curSeg := pathParts[pathPartsIdxCur]
+
+	if _, ok := hn.Children[curSeg]; !ok {
+		hn.Children[curSeg] = newMatchSequenceGeneratorNode()
+	}
+
+	child, _ := hn.Children[curSeg]
+	if curSeg == WildCardPathKey {
+		child.WildCard = true
+	}
+
+	return child.insertTemplate(pathParts, pathPartsIdxCur+1, httpMethod, methodData, markDuplicate)
+}
+
+func (hn *matchSequenceGeneratorNode) insertPath(pathParts []string, httpMethod string, methodData *methodData, markDuplicate bool) bool {
+	return hn.insertTemplate(pathParts, 0, httpMethod, methodData, markDuplicate)
+}
+
+func (hn *matchSequenceGeneratorNode) lookupPath(pathParts []string, pathPartsIdxCur int, httpMethod string, result *lookupResult) {
```

Comment:
Again, move "lookup" code to the bottom of the file with a big divider

Reply:
Done